map-low.go

```diff
@@ -105,11 +105,10 @@ func GetMapByID(id uint32) (*BPFMapLow, error) {
 func GetMapsIDsByName(name string) ([]uint32, error) {
 	bpfMapsIds := []uint32{}
 
-	startId := C.uint(0)
-	nextId := C.uint(0)
+	id := C.uint(0)
 
 	for {
-		retC := C.bpf_map_get_next_id(startId, &nextId)
+		retC := C.bpf_map_get_next_id(id, &id)
```

Comment:
I suppose passing the same object isn't a problem, since the libbpf helper only writes to it after reading it and when no error occurred. 👍🏼

But the problem you faced must be related to an error other than ENOENT. WDYT?

Reply:
**FWIK** the `GetMapsIDsByName` does not return an `error`. You can try locally by rollbacking the `GetMapsIDsByName` _(with StartId end EndId)_, and normally the test should fail because no IDs are returned. The error should be: `the test_hash_name map should be found`

Reply:
Sorry for the delay.

https://elixir.bootlin.com/linux/v6.5.7/source/kernel/bpf/syscall.c#L3778

`bpf_obj_get_next_id()` has four points of error return. We're considering only `ENOENT` in `GetMapsIDsByName()` what can cause a infinite loop when the error is one other.

In hindsight I assume that this wrapping is not one of the best choices, since we should abort it at any error returned but we don't provide a way to restart the fetching based on the last id successfully got.

My proposal is to create a proper wrapper getting only the next map id, using the same signature or similar like:

`GetMapNextID(startID uint32, nextID *uint32) error`

or

`GetMapNextID(startID uint32) (uint32, error)`

And convert `GetMapsIDsByName` to make use of GetMapNextID. In it consider all the types of errors, returning the filled slice of IDs (partial or not) and the error (whenever it occurs).

`GetMapsIDsByName(name string, startID uint32) ([]uint32, error)`

The GetMapsIDsByName caller would check for ENOENT, if it's, all maps IDs with that name would be returned. If it's a different kind, it's up to the user to try it again using the last ID from the []uint32. WDYT? Would you like to continue with these changes?

Reply:
I agree with you 👍 Handling all kinds of error could be nice and having a wrapper for `GetMapNexId` is a nice idea. But in the scenario I don't notice any error returned. Just the map is empty.

Reply:
> bpf_obj_get_next_id() has four points of error return. We're considering only ENOENT in GetMapsIDsByName() what can cause a infinite loop when the error is one other.

I don't think we are in an infinity loop if an error is different than ENOENT, because we wrap all other errors from `bpf_map_get_next_id`:

``` go
		retC := C.bpf_map_get_next_id(id, &id)
		errno := syscall.Errno(-retC)
		if retC < 0 {
			if errno == syscall.ENOENT {
				return bpfMapsIds, nil
			}

			return bpfMapsIds, fmt.Errorf("failed to get next map id: %w", errno)
		}
```

Reply:
For the GetMapNextId something like that could do the job? :
``` go

func GetMapNextID(startID uint32) (uint32, error) {
	retC := C.bpf_map_get_next_id(startID, &startID)
	errno := syscall.Errno(-retC)
	if retC < 0 {
		switch errno {
		case syscall.ENOENT:
			return startID, GetMapNextIDENOENTError{}
		case syscall.EINVAL:
			return startID, GetMapNextIDEINVALError{}
		case syscall.EPERM:
			return startID, GetMapNextIDEPERMError{}
		default:
			return startID, fmt.Errorf("failed to get next map id: %w", errno)
		}
	}

	return startID, nil
}
```

Reply:
> I don't think we are in an infinity loop if an error is different than ENOENT, because we wrap all other errors from bpf_map_get_next_id:

You're right. 👍🏼